elastic/kibana
# [SharedUX] Modify code editor theme definition and resolution implementation to account for color modes in EUI Theme

## Summary



Closes https://github.com/elastic/kibana/issues/202782

This PR reworks how custom themes used within the kibana code editor  for  the default visual look and ones specific to supported languages are defined to accomodate the upcoming visual refresh, the approach here leverages the `euiTheme` object value  returned from the `useEuiTheme` hook, now a single theme declaration is all that is required such that using either  the `colorMode `value  or the `euiTheme` from the provided `UseEUITheme` value it's possible to craft a theme that's in the context of kIbana, color mode aware and the editor would be able to resolve the appropriate colors depending on the user's color mode. 

This required some modification to monaco itself; now when defining languages if the  `CustomLanguageType` specification is being followed, a function that resolves to a standard monaco theme can be provided on the property `languageThemeResolver` which will be passed the `euiTheme` when registering this theme. It's worth mentioning that this can also be done manually by leveraging the custom method `registerLanguageThemeResolver` added  on the monaco editor object, like so 

```tsx
 monaco.editor.registerLanguageThemeResolver(LanguageID, languageThemeResolver);
``` 

However one should take note that when calling this method directly, the ID passed must correlate to a registered language ID, else the theme will not be available for use after Monaco is initialised, hence the theme name must equal an existing language ID if it's to be used for a specific language. 


## How to test

- Enable borealis, like so;
	- in your `kibana.dev.yml` file include the following config;
		```yml
		uiSettings.experimental.themeSwitcherEnabled: true
		```
	-  start kibana using the following command; `KBN_OPTIMIZER_THEMES="borealislight,borealisdark,v8light,v8dark" yarn start --run-examples`
- Tryout all downstream of the code editor to ascertain the code editor colors are as should be for both Amsterdam and Borealis; downstreams include; 
  - ES|QL editor, navigate to discover and click the "try ES|QL" button
  - Dev tools, on clicking the nav hamburger menu under the management menu group there's a menu item that links to all dev tools
  - Index Management use cases; navigate to stack management, under Index management select any existing index, then navigate to it's settings this should load up the code editor.
  - Saved Object use case; navigate to stack management, under saved objects attempt inspecting any saved object we'd be presented with the code editor  etc.
